### PR TITLE
Cleans up DPC inner and outer circuit instantiation/usage

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

This PR is a clean up of DPC inner and outer circuits. As such, no new functionalities are introduced and all existing tests remain passing. (Also no resampling is required)

### General Changes

- Adds a transaction benchmark in `snarkvm_dpc`

### Changes to DPC circuit implementations

- Adds `InnerPrivateVariables` and `OuterPrivateVariables`
- Renames `InnerCircuitVerifierInput` to `InnerPublicVariables` and begin using this struct in the inner circuit to reduce the number of clones and function argument inputs
- Renames `OuterCircuitVerifierInput` to `OuterPublicVariables` and begin using this struct in the inner circuit to reduce the number of clones and function argument inputs
- Updates `InnerCircuit` struct and methods to use `InnerPublicVariables/InnerPrivateVariables`, adds debug-only safety checks on program commitment and local data root
- Updates `OuterCircuit` struct and methods to use `OuterPublicVariables/OuterPrivateVariables`, adds debug-only safety checks on program commitment and local data root

### CODE REVIEW

**Comment by weikengchen** [comment]
How to run:

```
cd dpc
cargo bench --bench transaction
```

File: dpc/benches/transaction.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    let ledger = Ledger::<Testnet1Parameters, MemDb>::new(None, Block {
+        header: BlockHeader {
+            previous_block_hash: BlockHeaderHash([0u8; 32]),
+            merkle_root_hash: MerkleRootHash([0u8; 32]),
+            pedersen_merkle_root_hash: PedersenMerkleRootHash([0u8; 32]),
+            time: 0,
+            difficulty_target: 0xFFFF_FFFF_FFFF_FFFF_u64,
+            nonce: 0,
+            proof: ProofOfSuccinctWork([0u8; 771]),
+        },
```
Comment (ljedrz) [inline dpc/benches/transaction.rs:103]: note: this seems like a good candidate for a `Blockheader::default` or `BlockHeader::dummy` that could depend on the testnet

File: dpc/src/errors/dpc.rs [hunk 2]
```diff
@@ -81,7 +81,7 @@ pub enum DPCError {
     Message(String),
 
     #[error("missing inner snark proving parameters")]
```
Comment (ljedrz) [inline dpc/src/errors/dpc.rs:83]: shouldn't this message be adjusted too?

File: dpc/src/errors/dpc.rs [hunk 3]
```diff
@@ -90,7 +90,7 @@ pub enum DPCError {
     MissingNoopCircuit,
 
     #[error("missing outer snark proving parameters")]
```
Comment (ljedrz) [inline dpc/src/errors/dpc.rs:92]: ditto

**Review by ljedrz — APPROVED** [review]
LGTM :+1:

**Review by raychu86 — APPROVED** [review]

File: dpc/src/circuits/inner_private_variables.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        new_records_encryption_randomness: Vec<<C::AccountEncryptionScheme as EncryptionScheme>::Randomness>,
+        program_randomness: <C::ProgramCommitmentScheme as CommitmentScheme>::Randomness,
+        local_data_commitment_randomizers: Vec<<C::LocalDataCommitmentScheme as CommitmentScheme>::Randomness>,
+    ) -> Self {
+        assert_eq!(C::NUM_INPUT_RECORDS, input_records.len());
+        assert_eq!(C::NUM_INPUT_RECORDS, old_witnesses.len());
+        assert_eq!(C::NUM_INPUT_RECORDS, old_private_keys.len());
+        assert_eq!(C::NUM_OUTPUT_RECORDS, new_records.len());
+        assert_eq!(C::NUM_OUTPUT_RECORDS, new_records_encryption_randomness.len());
+
```
Comment (weikengchen) [inline dpc/src/circuits/inner_private_variables.rs:72]: Do you want to add a check that

```suggestion
        assert_eq!(C::NUM_TOTAL_RECORDS, local_data_commitment_randomizers.len());
        
```

**Review by weikengchen — APPROVED** [review]

File: dpc/benches/transaction.rs (hunk 1 above)
Comment (howardwu) [reply dpc/benches/transaction.rs:103]: Will follow up in a PR soon with this

**Pull request merged**